<!DOCTYPE html>
<html>
    
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <title> IsoPops | PCA + t-SNE </title>
        <link rel="stylesheet" href="main.css">
        <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">
        
    </head>
    <body>
        <nav class="navbar navbar-dark fixed-top flex-md-nowrap shadow navbar-bg">
            <a class="navbar-brand col-sm-3 col-md-9 mr-0" href="#"> IsoPops <span class="faded-white">Version 0.3.1 </span> </a>
        </nav>

        <div class="container-fluid main-container">
          <div class="row">
            <nav class="col-md-2 d-none d-md-block bg-dark sidebar sticky-top sticky-offset">
              <div class="sidebar-sticky sticky-top sticky-offset">
                <ul class="nav flex-column">
                  <li class="nav-item">
                    <a class="nav-link" href="../index.html"> Home </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="about.html"> About </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="install.html"> Install </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="walkthrough.html"> Walkthrough </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="https://github.com/kellycochran/IsoPops"> Source Code </a>
                  </li>
                </ul>

                <h6 class="sidebar-heading d-flex justify-content-between align-items-center px-3 mt-4 mb-1 text-muted">
                  <span class="sidebar-heading"> Docs </span>
                  <a class="d-flex align-items-center text-muted" href="#">
                    <span data-feather="plus-circle"></span>
                  </a>
                </h6>
                <ul class="nav flex-column mb-2">
                  <li class="nav-item">
                    <a class="nav-link" href="DB_objects.html"> Database Objects </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="walkthrough.html"> PCA + t-SNE <span class="sr-only"> (current) </span></a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="clustering.html"> Clustering </a>
                  </li>
                </ul>
                <h6 class="sidebar-heading d-flex justify-content-between align-items-center px-3 mt-4 mb-1 text-muted">
                  <span class="sidebar-heading"> Plot Functions </span>
                  <a class="d-flex align-items-center text-muted" href="#">
                    <span data-feather="plus-circle"></span>
                  </a>
                </h6>
                <ul class="nav flex-column mb-2">
                  <li class="nav-item">
                    <a class="nav-link" href="plots.html"> Standard Plots </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="jellyfish.html"> Jellyfish Plots </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="splice_heatmap.html"> Splice Heatmaps </a>
                  </li>
                </ul>
              </div>
            </nav>

            <main role="main" class="col-md-9 ml-sm-auto col-lg-10 px-4">

              <div class="">
                <div class="container mid-page">
                  <h1> Visualizing Isoform Populations with PCA or t-SNE </h1>
                  <br>
                    
                  <div class="centered-plot-pic">
                      <img class="card-img-top plot-pic-max-width" src="images/pca.png" alt="Example plot">
                  </div>
                    
                
                  <p> To interpret the isoform diversity in your data, it can be helpful to view the "spread" or "distribution" of transcripts for one or more genes in a low-dimensional setting. IsoPops can perform dimension reduction followed by visualization using either PCA or t-SNE, two popular unsupervised machine learning tools. The resulting plots display isoforms such that more similar isoforms are close to each other, while more divergent isoforms are farther away from one another. The process for generating these plots is:</p>
                      
                  <ol>
                    <li>Quantitatively represent transcript (or ORF) sequences as vectors via gene-agnostic, annotation-agnostic k-mer counting.</li>
                    <li>Run either PCA or t-SNE on these vectorized representations of isoforms.</li>
                    <li>Plot the PCA or t-SNE embeddings of the isoforms, in 2D or 3D.</li>
                  </ol>
                  
                  <p> These three steps are each performed with a single function call in IsoPops, as shown below: </p>
                    
                  <pre>
                    counts &lt;- get_kmer_counts(DB, genes = c("Ptprd", "Ptprf", "Ptprk", "Ptprs"))
                    pca &lt;- kmer_PCA(DB, counts)
                    plot_PCA(DB, pca)
                  </pre>
                    
                  <p>To plot in 3D, replace the last line with <code>plot_3D_PCA(DB, pca)</code>.
                  </p>
                    
                  <p>To run t-SNE instead:
                  </p>
                    
                  <pre>
                    counts &lt;- get_kmer_counts(DB, genes = c("Ptprd", "Ptprf", "Ptprk", "Ptprs"))
                    tsne &lt;- kmer_tSNE(DB, counts, iterations = 5000, perplexity = 40, dims = 2)
                    plot_tSNE(DB, tsne)
                  </pre>
                    
                  <p>To plot in 3D:
                  </p>
                    
                  <pre>
                    counts &lt;- get_kmer_counts(DB, genes = c("Ptprd", "Ptprf", "Ptprk", "Ptprs"))
                    tsne3D &lt;- kmer_tSNE(DB, counts, dims = 3)
                    plot_tSNE(DB, tsne3D, force_3D = T)
                  </pre>
                
                  <p>To generate these plots using ORF sequences instead of transcript sequences, simply add the argument <code>use_ORFs = T</code> to each function (including the plot functions).
                  </p>
                    
                  
                    
                </div>
                  
                <div class="footer text-center">
                  <hr>
                  IsoPops is licensed under the <a href="https://github.com/kellycochran/IsoPops/blob/master/LICENSE">GNU General Public License v3.0.</a> This site uses Twitter's <a href="https://getbootstrap.com">Bootstrap</a> framework.
                </div> 
              </div>
            </main>
          </div>
        </div>
        
        <!-- Bootstrap, jQuery -->
        <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
        <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>        
    </body>
</html>

