<!DOCTYPE html>
<html>
    
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <title> IsoPops | Standard Plots </title>
        <link rel="stylesheet" href="main.css">
        <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">
        
    </head>
    <body>
        <nav class="navbar navbar-dark fixed-top flex-md-nowrap shadow navbar-bg">
            <a class="navbar-brand col-sm-3 col-md-9 mr-0" href="#"> IsoPops <span class="faded-white">Version 0.3.1 </span> </a>
        </nav>

        <div class="container-fluid main-container">
          <div class="row">
            <nav class="col-md-2 d-none d-md-block bg-dark sidebar sticky-top sticky-offset">
              <div class="sidebar-sticky sticky-top sticky-offset">
                <ul class="nav flex-column">
                  <li class="nav-item">
                    <a class="nav-link" href="../index.html"> Home </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="about.html"> About </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="install.html"> Install </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="walkthrough.html"> Walkthrough </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="https://github.com/kellycochran/IsoPops"> Source Code </a>
                  </li>
                </ul>

                <h6 class="sidebar-heading d-flex justify-content-between align-items-center px-3 mt-4 mb-1 text-muted">
                  <span class="sidebar-heading"> Docs </span>
                  <a class="d-flex align-items-center text-muted" href="#">
                    <span data-feather="plus-circle"></span>
                  </a>
                </h6>
                <ul class="nav flex-column mb-2">
                  <li class="nav-item">
                    <a class="nav-link" href="DB_objects.html"> Database Objects </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="walkthrough.html"> PCA + t-SNE </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="clustering.html"> Clustering </a>
                  </li>
                </ul>
                <h6 class="sidebar-heading d-flex justify-content-between align-items-center px-3 mt-4 mb-1 text-muted">
                  <span class="sidebar-heading"> Plot Functions </span>
                  <a class="d-flex align-items-center text-muted" href="#">
                    <span data-feather="plus-circle"></span>
                  </a>
                </h6>
                <ul class="nav flex-column mb-2">
                  <li class="nav-item">
                    <a class="nav-link" href="plots.html"> Standard Plots <span class="sr-only"> (current) </span> </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="jellyfish.html"> Jellyfish Plots </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="splice_heatmap.html"> Splice Heatmaps </a>
                  </li>
                </ul>
              </div>
            </nav>

            <main role="main" class="col-md-9 ml-sm-auto col-lg-10 px-4">

              <div class="">
                <div class="container mid-page">
                  <h1> Standard Plots </h1>
                    
                  IsoPops implements the following suite of summary plots which may be useful for getting a sense of the isoform diversity in your data. All standard plot functions require the package ggplot2 and take in a processed Database object. These functions also allow for subsetting of the dataset by a list of genes, and most let you toggle between viewing summary statistics for transcripts and/or ORFs.
                    
                  <br><br>
                    
                  <h3> Isoform Length Distributions </h3>
                    <br>
                    <div class="centered-plot-pic">
                      <img class="card-img-top plot-pic-max-width" src="images/length_dist_plot.png" alt="Example plot">
                    </div>
                    <br>
                    
                    <code> plot_length_dist(database, use_ORFs = F, bins = 200, horiz_spread = 0.3, ...) </code>

                    <br><p>Generates a dot plot showing how transcripts/ORFs are distributed in length for each gene in the database.</p>

                  <h5> Arguments </h5>
                    <table class="table table-borderless table-sm">
                      <tbody>
                        <tr>
                          <th scope="row" style="width: 20%"><code>database</code></th>
                          <td>A compiled Database object.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>use_ORFs</code></th>
                          <td>Logical. Set to TRUE to use abundances from OrfDB instead of abundances from TranscriptDB. Note that OrfDB collapses isoforms with non-unique transcripts, so abundances may differ significantly.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>bins</code></th>
                          <td>The number of bins to use in segmenting the range of lengths over the entire dataset. This parameter determines vertical dot spread.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>genes_to_include</code></th>
                          <td>Vector of gene names to subset from the database. Default is to plot all genes in the database.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>horiz_spread</code></th>
                          <td>Numeric. This parameter determines horizontal dot spread across all the genes visualized.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>insert_title</code></th>
                          <td>String to customize the title of the plot.</td>
                        </tr>
                      </tbody>
                    </table>
                    
                  <h5> Returns </h5>
                    A length distribution dot plot constructed as a ggplot object.

                  <h5> Example Usage </h5>

                    <pre>
                      <span class="comment"> # database setup </span>
                      gene_ID_table &lt;- data.frame(ID = c("PB.1"), Name = c("Gene1"))
                      rawDB &lt;- compile_raw_db(transcript_file, abundance_file, gff_file, ORF_file)
                      DB &lt;- process_db(rawDB, gene_ID_table)
                      
                      plot_length_dist(DB)
                    </pre>
                    
                  <h5> Notes </h5>
                    Requires the <code>ggplot2</code> package.
                    
                    
                    
                    
                    <br><br><br><hr><br> 
                    
                    
                    
                  <h3> Treemap Plots </h3>
                    <br>
                    <div class="centered-plot-pic">
                      <img class="card-img-top plot-pic-max-width" src="images/treemap.png" alt="Example plot">
                    </div>
                    <br>
                    
                    <code> plot_treemap(database, use_ORFs = F, ...) </code>

                    <br><p>Generates a treemap plot showing how individual transcripts and genes account for abundance within the dataset as a whole</p>

                  <h5> Arguments </h5>
                    <table class="table table-borderless table-sm">
                      <tbody>
                        <tr>
                          <th scope="row" style="width: 20%"><code>database</code></th>
                          <td>A compiled Database object.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>use_ORFs</code></th>
                          <td>Logical. Set to TRUE to use abundances from OrfDB instead of abundances from TranscriptDB. Note that OrfDB collapses isoforms with non-unique transcripts, so abundances may differ significantly.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>genes_to_include</code></th>
                          <td>Vector of gene names to subset from the database. Default is to plot all genes in the database.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>insert_title</code></th>
                          <td>String to customize the title of the plot.</td>
                        </tr>
                      </tbody>
                    </table>
                    
                  <h5> Returns </h5>
                    A treemap plot constructed as a ggplot object.

                  <h5> Example Usage </h5>

                    <pre>
                      <span class="comment"> # database setup </span>
                      gene_ID_table &lt;- data.frame(ID = c("PB.1"), Name = c("Gene1"))
                      rawDB &lt;- compile_raw_db(transcript_file, abundance_file, gff_file, ORF_file)
                      DB &lt;- process_db(rawDB, gene_ID_table)
                      
                      plot_treemap(DB)
                    </pre>
                    
                  <h5> Notes </h5>
                    Requires the <code>ggplot2</code> package.

                                    
                    
                    
                    
                    <br><br><br><hr><br> 
                    
                    
                    
                  <h3> Exon-Abundance Distribution Plots </h3>
                    <br>
                    <div class="centered-plot-pic">
                      <img class="card-img-top plot-pic-max-width" src="images/ex_abund_dist_bar.png" alt="Example plot">
                    </div>
                    <br>
                    <div class="centered-plot-pic">
                      <img class="card-img-top plot-pic-max-width" src="images/ex_abund_dist_scatter.png" alt="Example plot">
                    </div>
                    <br>
                    
                    <code> plot_exon_dist(database,
                           sum_dist = T, bin_width = 0.02, ...) </code>

                    <br><p>Generates a bar plot showing the abundances of normalized exon counts for one or more genes. If each transcript is represented as the fraction of exons it contains out of the maximum number of exons found in a gene, this plot is merely a histogram of those representations, weighted by the read count for each transcript. Normalized exon percent is along the x-axis, and abundance is along the y-axis. If only one gene name is given, a second plot is generated where the x-axis is not normalized, instead showing the exon count of each transcript individually, and the y-axis is log-transformed. Jitter along the x-axis is added to improve visibility. </p>

                  <h5> Arguments </h5>
                    <table class="table table-borderless table-sm">
                      <tbody>
                        <tr>
                          <th scope="row" style="width: 20%"><code>database</code></th>
                          <td>A compiled Database object.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>genes_to_include</code></th>
                          <td>Vector of gene names to subset from the database. Default is to plot all genes in the database.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>sum_dist</code></th>
                          <td>Logical. If TRUE, the result is a histogram-like bar plot, where the x-axis is binned. Otherwise, individual isoforms are plotted as points, and the y-axis is log-transformed (single gene only).</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>bin_width</code></th>
                          <td>The histogram bin width, used only when multiple genes are input and the x-axis is the fraction of total exons per gene.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>insert_title</code></th>
                          <td>String to customize the title of the plot.</td>
                        </tr>
                      </tbody>
                    </table>
                    
                  <h5> Returns </h5>
                    A ggplot object.

                  <h5> Example Usage </h5>

                    <pre>
                      <span class="comment"> # database setup </span>
                      gene_ID_table &lt;- data.frame(ID = c("PB.1"), Name = c("Gene1"))
                      rawDB &lt;- compile_raw_db(transcript_file, abundance_file, gff_file, ORF_file)
                      DB &lt;- process_db(rawDB, gene_ID_table)
                      
                      plot_exon_dist(DB, sum_dist = T) <span class="comment"> # set to false for scatter plot </span>
                    </pre>
                    
                  <h5> Notes </h5>
                    Requires the <code>ggplot2</code> package.                 
                                    
                    
                    
                    
                    <br><br><br><hr><br> 
                    
                    
                    
                  <h3> Unique Isoforms/ORFs Barplots </h3>
                    <br>
                    <div class="centered-plot-pic">
                      <img class="card-img-top plot-pic-max-width" src="images/plot_counts.png" alt="Example plot">
                    </div>
                    <br>
                    
                    <code> plot_counts(database, use_log = F, use_counts = c("Isoforms", "ORFs")) </code>

                    <br><p>Generates a bar plot showing the number of unique isoform transcripts and unique ORFs for each gene.</p>

                  <h5> Arguments </h5>
                    <table class="table table-borderless table-sm">
                      <tbody>
                        <tr>
                          <th scope="row" style="width: 20%"><code>database</code></th>
                          <td>A compiled Database object.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>use_ORFs</code></th>
                          <td>Logical. Set to TRUE to use abundances from OrfDB instead of abundances from TranscriptDB. Note that OrfDB collapses isoforms with non-unique transcripts, so abundances may differ significantly.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>genes_to_include</code></th>
                          <td>Vector of gene names to subset from the database. Default is to plot all genes in the database.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>insert_title</code></th>
                          <td>String to customize the title of the plot.</td>
                        </tr>
                      </tbody>
                    </table>
                    
                  <h5> Returns </h5>
                    A ggplot object.

                  <h5> Example Usage </h5>

                    <pre>
                      <span class="comment"> # database setup </span>
                      gene_ID_table &lt;- data.frame(ID = c("PB.1"), Name = c("Gene1"))
                      rawDB &lt;- compile_raw_db(transcript_file, abundance_file, gff_file, ORF_file)
                      DB &lt;- process_db(rawDB, gene_ID_table)
                      
                      plot_N50_N75(DB)
                    </pre>
                    
                  <h5> Notes </h5>
                    Requires the <code>ggplot2</code> package.
                            
                                    
                    
                    
                    
                    <br><br><br><hr><br> 
                    
                    
                    
                  <h3> N50/N75 Barplots </h3>
                    <br>
                    <div class="centered-plot-pic">
                      <img class="card-img-top plot-pic-max-width" src="images/N50_N75.png" alt="Example plot">
                    </div>
                    <br>
                    
                    <code> plot_N50_N75(database, use_ORFs = F, ...) </code>

                    <br><p>Generates a bar plot showing the number of isoform transcripts and/or unique ORFs for each gene, using the thresholding concepts of N50 and N75. N50 refers to the minimum number of isoforms/ORFs needed to represent at least 50% of the abudance of a gene, while N75 refers to the minimum number of isoforms/ORFs needed to represent at least 75% of the abundance for a gene. This plot can help to identify which genes are dominated by very few isoforms. </p>

                  <h5> Arguments </h5>
                    <table class="table table-borderless table-sm">
                      <tbody>
                        <tr>
                          <th scope="row" style="width: 20%"><code>database</code></th>
                          <td>A compiled Database object.</td>
                        </tr>
                        <tr>
                          <th scope="row" style="width: 20%"><code>use_log</code></th>
                          <td>Logical. If true, y-axis is plotted on a log scale (base 2).</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>genes_to_include</code></th>
                          <td>Vector of gene names to subset from the database. Default is to plot all genes in the database.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>use_counts</code></th>
                          <td>One of both of the strings "Isoforms" and "ORFs", indicating which whould be included in the plot. Default is both and to give a warning if no ORF information is in the database.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>insert_title</code></th>
                          <td>String to customize the title of the plot.</td>
                        </tr>
                      </tbody>
                    </table>
                    
                  <h5> Returns </h5>
                    A ggplot object.

                  <h5> Example Usage </h5>

                    <pre>
                      <span class="comment"> # database setup </span>
                      gene_ID_table &lt;- data.frame(ID = c("PB.1"), Name = c("Gene1"))
                      rawDB &lt;- compile_raw_db(transcript_file, abundance_file, gff_file, ORF_file)
                      DB &lt;- process_db(rawDB, gene_ID_table)
                      
                      plot_isoform_orf_counts(DB)
                    </pre>
                    
                  <h5> Notes </h5>
                    Requires the <code>ggplot2</code> package.
                            
                                    
                    
                    
                    
                    <br><br><br><hr><br> 
                    
                    
                    
                  <h3> Shannon Diversity Index Plots </h3>
                    <br>
                    <div class="centered-plot-pic">
                      <img class="card-img-top plot-pic-max-width" src="images/shannon_index.png" alt="Example plot">
                    </div>
                    <br>
                    
                    <code> plot_Shannon_index(database, use_ORFs = F, ...) </code>

                    <br><p>Generates a plot showing the Shannon Index for isoform/ORF diversity on the x-axis, and each gene on the y-axis.</p>

                  <h5> Arguments </h5>
                    <table class="table table-borderless table-sm">
                      <tbody>
                        <tr>
                          <th scope="row" style="width: 20%"><code>database</code></th>
                          <td>A compiled Database object.</td>
                        </tr>
                        <tr>
                          <th scope="row" style="width: 20%"><code>use_ORFs</code></th>
                          <td>Logical. Set to TRUE to use abundances from OrfDB instead of abundances from TranscriptDB. Note that OrfDB collapses isoforms with non-unique transcripts, so abundances may differ significantly.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>genes_to_include</code></th>
                          <td>Vector of gene names to subset from the database. Default is to plot all genes in the database.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>insert_title</code></th>
                          <td>String to customize the title of the plot.</td>
                        </tr>
                      </tbody>
                    </table>
                    
                  <h5> Returns </h5>
                    A ggplot object.

                  <h5> Example Usage </h5>

                    <pre>
                      <span class="comment"> # database setup </span>
                      gene_ID_table &lt;- data.frame(ID = c("PB.1"), Name = c("Gene1"))
                      rawDB &lt;- compile_raw_db(transcript_file, abundance_file, gff_file, ORF_file)
                      DB &lt;- process_db(rawDB, gene_ID_table)
                      
                      plot_Shannon_index(DB)
                    </pre>
                    
                  <h5> Notes </h5>
                    Requires the <code>ggplot2</code> package.
                            
                                    
                    
                    
                    
                    <br><br><br><hr><br> 
                    
                    
                    
                  <h3> Exon Correlation Plots </h3>
                    <br>
                    <div class="centered-plot-pic">
                      <img class="card-img-top plot-pic-max-width" src="images/ex_corr_heatmap.png" alt="Example plot">
                    </div>
                    <br>
                    
                    <code> plot_exon_correlations(database, exon_filename, gene, weighted = T,
                                   exons_to_include = NULL, weights = NULL, plot_hist = F, symmetric = F) </code>

                    <br><p>Generates a 2D heatmap where each axis is the exons for a gene, and the values in the heatmap correspond to the correlation between the splicing events of pairs of exons. For example, the heatmap cell in row i and column j contains the pearson correlation of all the observed splicing inclusions and exclusions of exons i and j, according to the transcripts in the data. This plot can show which exons tend to either be included or spliced out together, for instance, and any exon pairs which may have mutually exclusive splicing patterns. Exon presence within a transcript is determined by literal string matching, so only full and completely correct matches between exon sequence and transcript sequence are considered.</p>

                  <h5> Arguments </h5>
                    <table class="table table-borderless table-sm">
                      <tbody>
                        <tr>
                          <th scope="row" style="width: 20%"><code>database</code></th>
                          <td>A compiled Database object.</td>
                        </tr>
                        <tr>
                          <th scope="row" style="width: 20%"><code>exon_filename</code></th>
                          <td>Path to a file in either FASTA or TSV format. If in FASTA format, the sequences are the annotated sequences for all exons in the gene, and the IDs are the exon names (will be displayed in the plot). The ID line must have format ">exonname". If in TSV format, There must be one column for exon names and one column for the exon sequence, tab-separated.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>gene</code></th>
                          <td>The desired gene to plot. Note that the plot will be generated only from exon matches to transcripts for the given gene, so no off-target exon matches are possible.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>weighted</code></th>
                          <td>Logical. If TRUE, transcript abundances will be taken into account when correlations are calculated (recommended).</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>exons_to_include</code></th>
                          <td>Vector of exon names to subset from the input file. Default is to include all exons in the inut file. This list is ordered; in other words, if you would like to rearrange the order of exon names on the axes of the heatmap, use this argument to do so.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>weights</code></th>
                          <td>A numeric vector specifying the weights to apply to each transcript for the given gene. Default is the number of full-length reads for the transcript.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>plot_hist</code></th>
                          <td>Logical. If TRUE, a histogram of all exon correlations across the gene is produced.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>symetric</code></th>
                          <td>Logical. If TRUE, both sides of the symmetric heatmap are shown.</td>
                        </tr>
                      </tbody>
                    </table>
                    
                  <h5> Returns </h5>
                    A ggplot object.

                  <h5> Example Usage </h5>

                    <pre>
                      <span class="comment"> # database setup </span>
                      gene_ID_table &lt;- data.frame(ID = c("PB.1"), Name = c("Gene1"))
                      rawDB &lt;- compile_raw_db(transcript_file, abundance_file, gff_file, ORF_file)
                      DB &lt;- process_db(rawDB, gene_ID_table)
                      
                      plot_exon_correlations(DB)
                    </pre>
                    
                  <h5> Notes </h5>
                    Requires the <code>ggplot2</code> package.
                    
                </div> 
                  
                  
                  
                  
                  
                  
                  
                  
                  
                  
                  
                  
                  
                  
                  
                <div class="footer text-center">
                  <hr>
                  IsoPops is licensed under the <a href="https://github.com/kellycochran/IsoPops/blob/master/LICENSE">GNU General Public License v3.0.</a> This site uses Twitter's <a href="https://getbootstrap.com">Bootstrap</a> framework.
                </div> 
              </div>
            </main>
          </div>
        </div>
        
        <!-- Bootstrap, jQuery -->
        <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
        <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>        
    </body>
</html>

