<!DOCTYPE html>
<html>
    
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <title> IsoPops | Jellyfish Plots </title>
        <link rel="stylesheet" href="main.css">
        <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">
        
    </head>
    <body>
        <nav class="navbar navbar-dark fixed-top flex-md-nowrap shadow navbar-bg">
            <a class="navbar-brand col-sm-3 col-md-9 mr-0" href="#"> IsoPops <span class="faded-white">Version 0.3.1 </span> </a>
        </nav>

        <div class="container-fluid main-container">
          <div class="row">
            <nav class="col-md-2 d-none d-md-block bg-dark sidebar sticky-top sticky-offset">
              <div class="sidebar-sticky sticky-top sticky-offset">
                <ul class="nav flex-column">
                  <li class="nav-item">
                    <a class="nav-link" href="../index.html"> Home </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="about.html"> About </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="install.html"> Install </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="walkthrough.html"> Walkthrough </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="https://github.com/kellycochran/IsoPops"> Source Code </a>
                  </li>
                </ul>

                <h6 class="sidebar-heading d-flex justify-content-between align-items-center px-3 mt-4 mb-1 text-muted">
                  <span class="sidebar-heading"> Docs </span>
                  <a class="d-flex align-items-center text-muted" href="#">
                    <span data-feather="plus-circle"></span>
                  </a>
                </h6>
                <ul class="nav flex-column mb-2">
                  <li class="nav-item">
                    <a class="nav-link" href="DB_objects.html"> Database Objects </a>
                  </li>
                </ul>
                <h6 class="sidebar-heading d-flex justify-content-between align-items-center px-3 mt-4 mb-1 text-muted">
                  <span class="sidebar-heading"> Plot Functions </span>
                  <a class="d-flex align-items-center text-muted" href="#">
                    <span data-feather="plus-circle"></span>
                  </a>
                </h6>
                <ul class="nav flex-column mb-2">
                  <li class="nav-item">
                    <a class="nav-link" href="plots.html"> Standard Plots </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="jellyfish.html"> Jellyfish Plots </a>
                  </li>
                  <li class="nav-item">
                    <a class="nav-link" href="splice_heatmap.html"> Splice Heatmaps <span class="sr-only"> (current) </span> </a>
                  </li>
                </ul>
              </div>
            </nav>

            <main role="main" class="col-md-9 ml-sm-auto col-lg-10 px-4">

              <div class="">
                <div class="container mid-page">
                  <h3> Splice Heatmaps </h3>
                    <br>
                    <div class="centered-plot-pic">
                      <img class="card-img-top" src="images/splicing.png" alt="Example plot">
                    </div>
                    <br>
                    
                    <code> plot_splicing(database, gene, max_isoforms = -1, zoom_in = c(0, 1), apply_max_to_splicing = F) </code>

                    <br><p>Generates a plot showing isoforms for one gene represented in genome-aligned format. Like a sashimi plot, splice junctions are represented as lines connecting splice donor and acceptor sites. Lines are colored according to the relative frequency of the splice junction occuring in the data. By default, the entire gene length and all isoforms are shown, but parameters allow for only the top X isoforms to be shown. In addition, an interval representing the portion of the gene body to zoom in on can be input, either as two fractions betwen 0 and 1, or as genomic coordinates. </p>

                  <h5> Arguments </h5>
                    <table class="table table-borderless table-sm">
                      <tbody>
                        <tr>
                          <th scope="row" style="width: 20%"><code>database</code></th>
                          <td>A compiled Database object.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>gene</code></th>
                          <td>Name of the gene to show.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>max_isoforms</code></th>
                          <td>Number of isoforms to plot. The top <code>[max_isoforms]</code> isoforms are chosen in decreasing order of abundance (most common isoforms first). </td>
                        </tr>
                        <tr>
                          <th scope="row"><code>zoom_in</code></th>
                          <td>Vector of two numbers, representing the interval of the gene length to show. If both numbers are between 0 and 1, inclusive, then this range is interpreted as fractions of the total gene length (e.g., <code>zoom_in = c(0.25, 0.75)</code> shows only the middle half of the gene). Otherwise, the numbers are interpreted as genomic coordinates for the chromosome that the gene is on.</td>
                        </tr>
                        <tr>
                          <th scope="row"><code>apply_max_to_splicing</code></th>
                          <td>Logical. If some isoforms are filtered out using <code>max_isoforms</code>, recalculate the splice junctions shown based on only the isoforms shown.</td>
                        </tr>
                      </tbody>
                    </table>
                    
                  <h5> Returns </h5>
                    A splice heatmap plot constructed as a ggplot object.

                  <h5> Example Usage </h5>

                    <pre>
                      <span class="comment"> # database setup </span>
                      gene_ID_table &lt;- data.frame(ID = c("PB.1"), Name = c("Gene1"))
                      rawDB &lt;- compile_raw_db(transcript_file, abundance_file, gff_file, ORF_file)
                      DB &lt;- process_db(rawDB, gene_ID_table)
                      
                      plot_splicing(DB, "Crb2")
                    plot_splicing(DB, "Crb2", max_isoforms = 5, zoom_in = c(0.9, 1))
                    plot_splicing(DB, "Crb2", max_isoforms = 5, zoom_in = c(64680000, 64694500))  <span class="comment"> # zoom in using genomic coordinates </span>
                    </pre>
                    
                  <h5> Notes </h5>
                    Requires the <code>ggplot2</code> package.
                    
                    This function utilizes R plotting functionality in unconventional ways, so it may not be as robust as other functions in this package. If you find a bug, please report through a Github issue or email.
                    
                </div>
                  
                <div class="footer text-center">
                  <hr>
                  IsoPops is licensed under the <a href="https://github.com/kellycochran/IsoPops/blob/master/LICENSE">GNU General Public License v3.0.</a> This site uses Twitter's <a href="https://getbootstrap.com">Bootstrap</a> framework.
                </div> 
              </div>
            </main>
          </div>
        </div>
        
        <!-- Bootstrap, jQuery -->
        <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
        <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>        
    </body>
</html>

